### PR TITLE
exdns: introduce GetFirstRR()

### PR DESCRIPTION
**Pull request by amery**
GetFirstRR() to get the first of a type in msg.Ns or msg.Extra
IsDescendent() to validate delegations

### CODE REVIEW

File: pkg/exdns/exdns.go [hunk 1]
```diff
@@ -39,3 +40,19 @@ func Decanonize(qName string) string {
...
+	case !ok:
+		return false
+	case ancestor == ".":
+		return true
+	case s == "", s[len(s)-1] == '.':
+		return true
+	default:
+		return false
+	}
+}
```
Comment (karasz) [inline pkg/exdns/exdns.go:58]: isn't this an alternative and a bit more fragile implementation of [dns.IsSubdomain(parent, child string) bool )](https://pkg.go.dev/github.com/miekg/dns#IsSubDomain) ?

If yes (and as far as I can see it is.) please remove it.

**Review by karasz — CHANGES_REQUESTED** [review]
LGTM with the commented caveat. `GetFirstRR` is ok.

File: pkg/exdns/exdns.go (hunk 1 above)
Comment (amery) [reply pkg/exdns/exdns.go:58]: I had the impression `IsSubdomain()` only accepts direct descendants, am I wrong?

Comment (karasz) [reply pkg/exdns/exdns.go:58]: I am not sure what you mean by "direct descendants" but `IsSubdomain()` will check if `ancestor` can be removed from `child`. So will return true for `(example.com.,test1.example.com)` and `(example.com, test2.test1.example.com)` and `(test1.example.com, test2.test1.example.com)` but false for `(example.net, example.com)` or `(example.net, test1.anotherexample.net)`

**Comment by amery** [comment]
@karasz IsDescendant() dropped  in favour of `dns.IsSubDomain()`

**Review by karasz — APPROVED** [review]
LGTM

**Comment by amery** [comment]
`v0.7.23` tagged

**Pull request merged**